File: asgn2.2/questionmc.cpp
/** ---------------------------------------------
*
* @file		questionmc.cpp
* @author	Nick McCollum
* @version	1.0
*
* CS-202 - Assignment 2 - Inheritance & Polymorphism
*		   QuestionMC class implementation
*
* -------------------------------------------- */

#include <iostream>
#include <sstream>
#include "questionmc.h"

using namespace std;

// ----------------------------------------------
QuestionMC::QuestionMC(string questionText = "", string answerText = "", string answerChoices = "")
	: Question(questionText, answerText)
{
	this->questionText = questionText;
	this->answerText = answerText;
	this->answerChoices = answerChoices;
	correct = false;
}

// ----------------------------------------------
void QuestionMC::showQuestion() {
	vector<string> choices;
	parseChoices(choices);

	cout << questionText;
	cout << "?";

	char choiceLetter = 'a';
	for (unsigned int i = 0; i < choices.size(); i++) {
		cout << "\n   " << choiceLetter << ") ";
		cout << choices[i];
		choiceLetter++;
	}
}

void QuestionMC::showAnswer() {
	cout << answerText;
}

// ----------------------------------------------
bool QuestionMC::checkAnswer(string givenAnswer) {
	string actualAnswer = answerText;

	// Remove whitespace and convert to lower
	removeWhitespace(actualAnswer);
	removeWhitespace(givenAnswer);
	toLowerCase(actualAnswer);
	toLowerCase(givenAnswer);

	if (givenAnswer[0] != actualAnswer[0]) {
		return false;
	}

	markCorrect();
	return true;
}

// ----------------------------------------------
bool QuestionMC::parseChoices(vector<string> &choices) {
	istringstream choiceStream(answerChoices);
	string token;

	while (!choiceStream.eof()) {
		getline(choiceStream, token, ':');
		choices.push_back(token);
	}

	if (choiceStream.fail()) {
		cerr << "ERROR: Failed reading multiple choice answers.\n";
		choiceStream.clear();
		return false;
	}

	return true;
}

File: Quiz/QuestionMC.java
/**
 * CS-202 - Assignment 5 - Java Classes
 * Multiple Choice question object used by Quiz.
 * Displays MC question and checks the user's answer.
 * 
 * @author Nick McCollum
 * @version 1.0
 *
 */

public class QuestionMC extends Question {

	public QuestionMC(String question, String answer, String choices) {
		super(question, answer);
		this.question = question;
		this.answer = answer;
		this.choices = choices;
		correct = false;
	}

	// ------------------------------------------------------------------
	/**
	 * Displays the MC question text in the console.
	 */
	public void showQuestion() {
		String[] choiceArray = parseChoices(choices);
		System.out.print(question + "?");
		
		char choiceLetter = 'a';
		for (String choice : choiceArray)  {
			System.out.print("\n\t" + choiceLetter + ") " + choice);
			choiceLetter++;
		}
			
	}
	
	// ------------------------------------------------------------------
	/**
	 * Checks if the answer provided by the user was correct.
	 * @param givenAnswer	The answer entered by the user.
	 * @return boolean	True if the answer was correct, false if not.
	 */
	public boolean checkAnswer(String givenAnswer) {
		givenAnswer = removeWhitespace(givenAnswer.toLowerCase());
		String actualAnswer = removeWhitespace(answer.toLowerCase());
		
		if (givenAnswer.charAt(0) != actualAnswer.charAt(0)) {
			return false;
		}
		
		markCorrect();
		return true;
	}
	
	protected String choices;
	
	protected String[] parseChoices(String choices) {
		return choices.split(":");
	}
}
